Flatpak Package Build System & Support

someone13574 opened the pull request:
This PR adds support for building Flatpak packages, as well as the necessary metainfo/desktop entries needed by Zed and fixing the terminal to function within a Flatpak. For to moment I've forgone adding screenshots and release notes to the appstream manifest.

In order to fix the terminal, the Flatpak also compiles and bundles [host-spawn](https://github.com/1player/host-spawn) which allows it to spawn the shell process outside of the sandbox. This is used over the normal [flatpak-spawn](https://man7.org/linux/man-pages/man1/flatpak-spawn.1.html) due to [problems](https://github.com/flatpak/flatpak-xdg-utils/issues/57) which it has with running processes which require a tty/pty.

Useful Docs:

- [Flatpak](https://docs.flatpak.org/en/latest/)
- [Appstream](https://www.freedesktop.org/software/appstream/docs/chap-CatalogData.html#spec-asxml-tags)
- [Desktop Entries](https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html)
- [Desktop Menu Categories](https://www.freedesktop.org/wiki/Specifications/menu-spec/)

Screenshot:

![Screenshot from 2024-04-24 22-53-02](https://github.com/zed-industries/zed/assets/81528246/43c6a2aa-b245-40aa-b206-b71a91c95787)

Release Notes:

- N/A

someone13574 committed:
Flatpak building

maxdeviant commented on script/flatpak/generated-sources.json:
Does this file need to be in Git?

someone13574 replied:
No. I'll remove it.

someone13574 committed:
Delete generated-sources.json

someone13574 commented:
Seems like the terminal is having problems with flatpaks sandboxing. Marking as a draft until I fix it.

someone13574 committed:
Fix terminal in flatpak sandbox
Fixes the flatpak terminal by prefixing the shell command with [host-spawn](https://github.com/1player/host-spawn). This runs the shell on the host-system, giving access to the programs which a user would expect. Note that this breaks through the sandbox. host-spawn is used over flatpak-spawn, the command built into flatpaks, because it allocates a pty for the spawned process, fixing issues with sudo and other commands.

someone13574 committed:
Update appstream description

someone13574 committed:
Give flatpak access to gpu acceleration

someone13574 commented:
Seems like git-blame and dev-extension installation is also failing due to trying to spawn a process on the host. These will need to run through flatpak-spawn or host-spawn as well.

someone13574 committed:
Move host-spawn module infront of zed module
This avoids recompiling it when zed needs to recompile

someone13574 committed:
Remove debugging println

Rotonen commented:
VS Code had the same kind of flapping about when it started to get packaged and distributed as a Flatpak (things got fixed one thing at a time over versions, but also kept breaking when things changed).

I suggest to consider adding (at least down the line, if a regression actually arises) a separate smoketest build for the Flatpak package as that stuff is fragile and the whole landscape around it is moving.

someone13574 commented:
> VS Code had the same kind of flapping about when it started to get packaged and distributed as a Flatpak (things got fixed one thing at a time over versions, but also kept breaking when things changed). Disclaimer: I am only a user of VS Code.
> 
> I suggest to consider adding (at least down the line, if a regression actually arises) a separate smoketest build for the Flatpak package as that stuff is fragile and the whole landscape around it is moving.

The problems that would arise wouldn’t really be possible to test just with a build, as it isn’t the build failing but rather the editor trying to spawn a process at runtime, which it cannot locate because it’s in the sandbox. If a problem does occur, it will be runtime only and will likely be from a new feature being added which doesn’t call host-spawn. This means that you cannot really write a test for it since the feature doesn’t exist at the moment. As for test builds, that already happens as clippy is configured with the —all-features flag. This doesn’t actually build the flatpak, but it does compile the code used for the separate spawning logic. Adding a test which actually compiles in the flatpak environment would be good though.

Rotonen commented:
Please excuse the loose language. Build == GHA Workflow. Collides with static language build, thus causing the confusion above.

Exactly, as the problems are runtime specific, one would need a novel trivial end to end test run (== build == new check in the per commit / PR build status box) for the resulting Flatpak via `flatpak run`. Orchestrating that is probably non-trivial on a CI as one would need to figure out an UI testing framework which would work here in an `xvfb` context or somesuch. All of which is wildly out of scope here for this PR.

We do not ultimately disagree here. And we do not need to continue the discussion further at this point in time - I am already draining resources further than intended. Willing to volunteer down the line in case the direction becomes desirable: simply poke me with an @ on an issue or a PR and I'll figure signing the CLA out.

mikayla-maki commented:
@someone13574 Since we're going to need to have custom for flatpak in many different locations, what do you think about building a little `zed-process` library that wraps a `Command` call and contains the flatpak `#[cfg]` calls? Then we can route these other uses (git, LSPs, etc.) through that interface.

someone13574 committed:
Add process wrapper crate and fix git blame

someone13574 commented on assets/packaging/dev.zed.Zed.svg:
This needs a background so that it doesn’t look terrible on white backgrounds

someone13574 committed:
Replace git checkout with git reset in flatpak-cargo-generator.py
This fixes an issue where the checkout would fail due to it checking out to a point where one of the files no longer existed in tree-sitter-cpp

someone13574 committed:
Merge branch 'main' into flatpak

someone13574 committed:
Wrap the windows CommandExt for process::Command

someone13574 committed:
Fix cargo machete
process and terminal need to be a dependency of zed even though they are unused because the flatpak compile flag must be passed workspace wide.

someone13574 committed:
Use existing app icon

someone13574 commented on script/flatpak/host-spawn/sources.json:
```diff
@@ -0,0 +1,35 @@
+[
+    {
+        "type": "archive",
+        "url": "https://github.com/someone13574/host-spawn/archive/refs/tags/temp.zip",
```
Temporary: DO NOT MERGE.

This should wait until [https://github.com/1player/host-spawn/pull/36](this PR) is merged upstream.

someone13574 committed:
Use process wrapper for extension building

someone13574 committed:
Add smol async to process wrapper

someone13574 committed:
Replace more process::Command and smol::Command calls

someone13574 committed:
Document the process wrapper

someone13574 committed:
Replace stray Command::new with process::Process in node_runtime.rs

someone13574 committed:
Fix terminal process_name

someone13574 committed:
Merge remote-tracking branch 'origin/main' into flatpak

someone13574 committed:
Improve flatpak README.md

mikayla-maki commented:
@someone13574 Since this PR has a comment saying 'do not merge', I'm going to reset it to a draft. When you're ready to merge let's come back to it :D

someone13574 commented:
> @someone13574 Since this PR has a comment saying 'do not merge', I'm going to reset it to a draft. When you're ready to merge let's come back to it :D

Sounds good. I was just thinking that merging while bundling a fork owned by me wouldn't be the best idea, even if the the only change from upstream is pretty trivial. The host-spawn maintainer was open to the idea, so once the PR gets merged upstream I'll switch over to that and then mark as ready. I don't foresee any other major changes at this moment other using upstream for host-spawn.

mikayla-maki commented:
Sounds good, I'll give this a code review at some point :)

mikayla-maki commented:
Thank you for working on this!

I took a look through this PR and I don't think we're interested in supporting flatpak until the official `flatpak-spawn` command is able to support our terminal natively. I'd rather not depend on a smaller project at such a critical part of the interface.  It seems there's an alternative solution, involving smarter integration with [the process spawned](https://github.com/flatpak/flatpak/issues/3285#issuecomment-1815265041), that we'd be more interested in.

Either way though, I'm going to close this PR for now. We'll be releasing `.deb` builds at some point soon, if it's not possible to wrap those builds for flatpak somehow, please file an issue and we can re-visit this once our Linux builds are more stable :)

someone13574 commented:
> Thank you for working on this!
> 
> I took a look through this PR and I don't think we're interested in supporting flatpak until the official `flatpak-spawn` command is able to support our terminal natively. I'd rather not depend on a smaller project at such a critical part of the interface. It seems there's an alternative solution, involving smarter integration with [the process spawned](https://github.com/flatpak/flatpak/issues/3285#issuecomment-1815265041), that we'd be more interested in.
> 
> Either way though, I'm going to close this PR for now so we can focus on stabilizing Linux builds. We'll be releasing `.deb` builds at some point soon, if it's not possible to wrap those builds for flatpak somehow, please file an issue and we can re-visit this once our Linux builds are more mature :)

I believe that the [openpty](https://github.com/alacritty/alacritty/blob/ce800bfde2a2121a830c2b0854749875b3aebf79/alacritty_terminal/src/tty/unix.rs#L185) command currently used by `alacritty_terminal` [already sets O_NOCTTY](https://github.com/sunfishcode/rustix-openpty/blob/5ae72174e1672dd4b52fb994d126e366fd28b0e6/src/lib.rs#L115). The real problem is that one needs to use the `--forward-fd` argument of flatpak-spawn, but since the pty is opened from inside alacritty, you don't know what they are at the time of constructing the flatpak-spawn command. Host-spawn was getting around this by just opening their own pty, but I agree that relying on a small project like that isn't a good solution.

As I currently see it, there are really only two ways which flatpak support could work in the future, if it is wanted. First would be building a small rust binary target into Zed which does what host-spawn does, calling the underlying [org.freedesktop.Flatpak.Development](https://docs.flatpak.org/en/latest/libflatpak-api-reference.html#gdbus-org.freedesktop.Flatpak.Development) dbus interface and allocating a second pty. The only difference would be that the code would then be hosted in the Zed repository instead of being external, and it would be written in rust like the rest of the project. The second way would be dropping alacritty partially or fully, which isn't really feasible due to the extent of it within Zed's terminal.

As for wrapping `.deb` builds for flatpak, that would only encounter the same problem of any processes designed to be run on the host being run in the sandbox. I understand the reasoning behind waiting for Zed on linux to become more mature before looking into pursuing this further.

someone13574 commented:
I’ve gotten a [change merged](https://github.com/alacritty/alacritty/commit/ed3fac8a033acf10293712339185920849d45a0a) in alacritty which allows us to open the pty ourselves instead of alacritty doing it, which means the file descriptor argument of flatpak-spawn can be used instead of relying on host-spawn. This should allow host-spawn to be dropped entirely if flatpak support is perused in the future.
